designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,294 @@
+- Feature Name: stan-output-formats
+- Start Date: 2022-01-11
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+In order to monitor the progress of CmdStan's inference algorithms,
+the outputs from CmdStan must be available to downstream readers on a streaming basis.
+This design provides an alternative to the [Stan CSV file format](https://mc-stan.org/docs/cmdstan-guide/stan_csv.html)
+for outputs of the Stan inference algorithms which are exposed by the `stan::services` layer.
+
+
+# Motivation
+[motivation]: #motivation
+
+The [Stan CSV file format](https://mc-stan.org/docs/cmdstan-guide/stan_csv.html) contains a record
+of the inference algorithm outputs.  This is the output format used by CmdStan, and therefore by the
+CmdStan wrapper interfaces CmdStanPy and CmdStanR  This format is limited and limiting for several reasons.
+
+* While the plain-text format is human readable, the drawbacks are that conversion to/from binary
+is expensive and may lose precision, unless default settings are overridden, which in turn will
+result in large output file size.  The plain text format is untyped, therefore we cannot distinguish
+between integer, real, or complex numbers.
+
+* The HMC sampler uses comment rows following the header row to report the stepsize, metric type, and metric
+and a final set of comment rows to report sampler timing information.
+This requires writing ad-hoc parsers to recover this information and precludes the use of many fast CSV parser libraries.
+
+
+* The Stan CSV file must necessarily use a single table to hold both the inference algorithm state as well as
+the model outputs produced by the model class's `write_array`.
+By convention, the initial columns contain inference algorithm state.
+The fist column is the estimated joint log probability `lp__`.
+Depending on the inference algorithms, there are zero or more columns whose names end in `__`.
+For HMC these are `accept_stat__`,`stepsize__`,`treedepth__`,`n_leapfrog__`,`divergent__`,`energy__`,
+for ADVI these are `log_p__` and `log_g__`.
+The optimization algorithms don't report any state information beyond `lp__`.
+
+* Because each row of the CSV file corresponds to one draw, there is no way to monitor the inner workings
+of the inference algorithm, e.g., leapfrog steps or gradient trajectories.
+
+* Although we can now run multiple chains in a single process, it is still necessary to produce per-chain CSV files,
+with the chain id baked into the filename.   A cleaner solution would be to combine all outputs in a single file.
```

Comment:
This would get pretty expensive from an i/o side because then we have multiple threads writing to a single file. For any output format I think we would need to use something clever to make sure writing for one chain would not go over and start writing where another chain just wrote in the file.

Reply:
good point - will add more discussion.